### DISCUSSION TITLE
How to unload ControlNet from the postprocess() routine of another extension?

### DISCUSSION THREAD

**Discussion by Elevory** [Mikubill/sd-webui-controlnet#726] [Q&A]
Hi,

After processing img2img with ControlNet, I would like to disable/unhook the CN modules and run another img2img operation using the result of the first operation as the new initial image. What is the correct way to go about this?

When CN is enabled, it does not seem to consider changes to the `p.initial_image` variable when `process_images()` is fired from the `postprocess()` routine of another extension. It just re-uses the first image. If I could update the image, then perhaps I wouldn't need to disable CN outright. I would be happy with either solution.

In case it's relevant, my extension runs before CN due to alphabetical priority.

Thanks. Any thoughts would be appreciated.

**Comment by ljleb** [comment]
I am not aware of the context, if you are comfortable sharing a link to your extension or explaining what you are trying to do, it would help with understanding your needs.

To update the control maps, you need to update the tensors at `cn_script.latest_network.control_params[i].hint_cond`. Note however that this may not play well with other extensions that also replace these maps. (see for example the multidifusion upscaler extension)

I am not sure, but I think you may want to look into one of these alternatives:

- If your script intends to generate/schedule multiple generations, you may want to look into making it a selectable script. By doing this, you would be able to call `process_images` as many times as you like.
- If you want to create some sort of alternative hires fix pass, you could monkey patch `processing.process_images_inner` and call the original function as many time as you need while passing the same `p` as argument.

Hope this helps. Let me know if your situation has requirements that are incompatible with these suggestions.

**Comment by Elevory** [comment]
Thank you, @ljleb! Great info.

I think the monkey patching approach may suit my needs. For context, I'm working on an update to Unprompted's [Zoom Enhance](https://www.reddit.com/r/StableDiffusion/comments/11pyiro/new_feature_zoom_enhance_for_the_a111_webui/) feature, which upscales a portion of an image and pastes it back onto the original. It is a "partial hires fix" that can be used to improve small details like faces or hands.

Anyhow, I was able to call `process_images_inner()` with a `p` object containing the right initial image. However, it still seems to re-apply Control Net to the final result. Does that mean the monkey patch wasn't implemented correctly? I am fairly new to the concept and am trying to determine if any other changes are needed to restore "normal" img2img inference.

Either way, I don't necessarily mind running CN again, but I think I would also have to run the pre-processors on the new initial image. My feeling was that disabling CN outright would be the simpler approach, but I'm not sure how to go about updating the tensors for `cn_script.latest_network.control_params[i].hint_cond`. Is there a code example I could look at?

Thanks again.

**Reply by ljleb** [reply]
To disable the controlnet extension after the first pass, you can use this code: (extracted from the [external code api wiki](https://github.com/Mikubill/sd-webui-controlnet/wiki/API#external-code-api))

```python
import importlib
external_code = importlib.import_module('extensions.sd-webui-controlnet.scripts.external_code', 'external_code')

external_code.update_cn_script_in_processing(p, []) # update the controlnet script args with a list of 0 units
```

If you want to enable controlnet for the first pass but disable during the second, run this between the 2 calls to `process_images`.

**Reply by Elevory** [reply]
Hmm, I must be doing something wrong. Even with that code, the second img2img pass seems to be affected by ControlNet:

![image](https://user-images.githubusercontent.com/29494707/231315358-3b3be5ac-c846-4b50-b201-a15ef34639da.png)

If I manually disable ControlNet and run img2img with the exact same settings, I get this:

![image](https://user-images.githubusercontent.com/29494707/231315524-0a230f46-4639-43d2-8947-d6ef4cc682b6.png)

In fact, it seems the face becomes more distorted with the number of ControlNet units I have active.

What am I missing?

**Reply by ljleb** [reply]
Is your code located somewhere? It would be easier if I could glaze at the code. It's hard for me to say as it could be anything AFAIK.

**Reply by Elevory** [reply]
Thanks for offering to take a look. Let me clean up the script (so many redundant bits after trying to implement this like 5 different ways, haha...) and push to my repo. Will provide a link soon.
